src/librustc_mir/borrow_check/mod.rs

```diff
@@ -1776,6 +1777,22 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
             }
 
             if let Some((prefix, mpi)) = shortest_uninit_seen {
+                // Check for a reassignment into a uninitialized field of a union (for example,
+                // after a move out). In this case, do not report a error here. There is an
+                // exception, if this is the first assignment into the union (that is, there is
+                // no move out from an earlier location) then this is an attempt at initialization
+                // of the union - we should error in that case.
+                let tcx = this.infcx.tcx;
+                if let ty::TyKind::Adt(def, _) = base.ty(this.mir, tcx).to_ty(tcx).sty {
+                    let moved_before_this = this.move_data.path_map[mpi].iter().any(|moi| {
+                        this.move_data.moves[*moi].source < context.loc
```

Comment:
.... is the `<` comparison here always going to reflect the control flow... ?

I would have guessed that our `PartialOrd`/`Ord` implementation is based around the numbers assigned to basic blocks, which I suspect often but *not always* correspond to control flow order...

Reply:
Yeah, I wasn't too sure about this either. I couldn't think of anything better at the time. I'll spend some time thinking about a way I could break this and see if I can't work out a more robust check.

Reply:
I've pushed a commit that changes this to look at the predecessors for a location - as I understand it, this should now correctly consider the cases where a straight comparison of the location would have failed.